File: Diary10/app/src/main/java/com/polarstation/diary10/activity/EditAccountActivity.java
package com.polarstation.diary10.activity;

import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import gun0912.tedkeyboardobserver.TedKeyboardObserver;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.view.View;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.storage.FirebaseStorage;
import com.polarstation.diary10.R;
import com.polarstation.diary10.databinding.ActivityEditAccountBinding;
import com.polarstation.diary10.util.NetworkStatus;

import java.util.HashMap;
import java.util.Map;

import static com.polarstation.diary10.util.NetworkStatus.TYPE_CONNECTED;

public class EditAccountActivity extends BaseActivity implements View.OnClickListener{
    private ActivityEditAccountBinding binding;
    private FirebaseStorage strInstance;
    private FirebaseDatabase dbInstance;
    private String uid;
    private String imageUrl = "";
    private Uri imageUri;
    private int netStat;

    public static final String NAME_KEY = "nameKey";
    public static final String COMMENT_KEY = "commentKey";
    public static final String URL_KEY = "urlKey";
    public static final String URI_KEY = "uriKey";
    public static final int PICK_FROM_ALBUM_CODE = 100;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_edit_account);

        Intent intent = getIntent();
        if(intent != null) {
            String urlFromAccount =  intent.getStringExtra(URL_KEY);
            String name = intent.getStringExtra(NAME_KEY);
            String comment = intent.getStringExtra(COMMENT_KEY);

            Glide.with(getBaseContext())
                    .load(urlFromAccount)
                    .apply(new RequestOptions().circleCrop())
                    .into(binding.editAccountActivityProfileImageView);
            binding.editAccountActivityNameEditText.setText(name);
            binding.editAccountActivityCommentEditText.setText(comment);
        }

        strInstance = FirebaseStorage.getInstance();
        dbInstance = FirebaseDatabase.getInstance();
        uid = FirebaseAuth.getInstance().getCurrentUser().getUid();

        binding.editAccountActivityProfileImageView.setOnClickListener(this);
        binding.editAccountActivityCloseButton.setOnClickListener(this);
        binding.editAccountActivitySaveButton.setOnClickListener(this);

        // 키보드 올라오거나 내려갈때 이벤트
        new TedKeyboardObserver(this).listen(isShow-> {
            if(!isShow && binding.editAccountActivityNameEditText.isFocused())
                binding.editAccountActivityNameEditText.clearFocus();
            else if(!isShow && binding.editAccountActivityCommentEditText.isFocused())
                binding.editAccountActivityCommentEditText.clearFocus();
        });
    }

    @Override
    public void onClick(View view) {
        switch(view.getId()){
            case R.id.editAccountActivity_closeButton:
                finish();
                break;
            case R.id.editAccountActivity_saveButton:
                netStat = NetworkStatus.getConnectivityStatus(getApplicationContext());
                if(netStat == TYPE_CONNECTED) {
                    String name = String.valueOf(binding.editAccountActivityNameEditText.getText());
                    if (name.equals(""))
                        Toast.makeText(this, getString(R.string.write_name), Toast.LENGTH_SHORT).show();
                    else {
                        Intent resultIntent = new Intent();
                        String comment = String.valueOf(binding.editAccountActivityCommentEditText.getText());

                        Map<String, Object> map = new HashMap<>();
                        map.put(getString(R.string.fdb_user_name), name);
                        dbInstance.getReference().child(getString(R.string.fdb_users)).child(uid).updateChildren(map);
                        map.clear();
                        map.put(getString(R.string.fdb_comment), comment);
                        dbInstance.getReference().child(getString(R.string.fdb_users)).child(uid).updateChildren(map);

                        resultIntent.putExtra(NAME_KEY, name);
                        resultIntent.putExtra(COMMENT_KEY, comment);

                        if (imageUri != null)
                            updateProfileImage(imageUri, resultIntent);
                        else {
                            binding.editAccountActivityProgressBar.setVisibility(View.INVISIBLE);
                            resultIntent.putExtra(URI_KEY, "");
                            setResult(RESULT_OK, resultIntent);
                            finish();
                        }
                    }
                }else Toast.makeText(getBaseContext(), getString(R.string.network_not_connected), Toast.LENGTH_SHORT).show();
                break;
            case R.id.editAccountActivity_profileImageView:
                Intent requestIntent = new Intent(Intent.ACTION_PICK);
                requestIntent.setType(MediaStore.Images.Media.CONTENT_TYPE);
                startActivityForResult(requestIntent, PICK_FROM_ALBUM_CODE);
                break;
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if(requestCode == PICK_FROM_ALBUM_CODE && resultCode == Activity.RESULT_OK){
            imageUri = data.getData();

            Glide.with(getBaseContext())
                    .load(imageUri)
                    .apply(new RequestOptions().circleCrop())
                    .into(binding.editAccountActivityProfileImageView);
        }
    }

    private void setViewWhenUploading(){
        binding.editAccountActivityProgressLayout.setVisibility(View.VISIBLE);
        binding.editAccountActivityConstraintLayout.setEnabled(false);
        binding.editAccountActivityCloseButton.setEnabled(false);
        binding.editAccountActivitySaveButton.setEnabled(false);
        binding.editAccountActivityProfileImageView.setEnabled(false);
        binding.editAccountActivityCommentEditText.setEnabled(false);
        binding.editAccountActivityNameEditText.setEnabled(false);
    }

    private void updateProfileImage(Uri imageUri, Intent resultIntent){
        setViewWhenUploading();
        strInstance.getReference().child(getString(R.string.fstr_user_images)).child(uid).putFile(imageUri)
                .addOnCompleteListener(task -> {
                    strInstance.getReference().child(getString(R.string.fstr_user_images)).child(uid).getDownloadUrl().addOnSuccessListener(uri -> {
                        imageUrl = String.valueOf(uri);
                        resultIntent.putExtra(URI_KEY, String.valueOf(imageUrl));

                        Map<String, Object> stringObjectMap = new HashMap<>();
                        stringObjectMap.put(getString(R.string.fdb_profile_image_url), imageUrl);
                        dbInstance.getReference().child(getString(R.string.fdb_users)).child(uid).updateChildren(stringObjectMap)
                        .addOnSuccessListener( aVoid -> {
                            setResult(RESULT_OK, resultIntent);
                            finish();
                        });
                    });

                });
    }
}

File: Diary10/app/src/main/java/com/polarstation/diary10/activity/EditAccountActivity.kt
package com.polarstation.diary10.activity

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.UserProfileChangeRequest
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.storage.FirebaseStorage
import com.polarstation.diary10.R
import com.polarstation.diary10.databinding.ActivityEditAccountBinding
import com.polarstation.diary10.fragment.*
import com.polarstation.diary10.util.FontUtil
import com.polarstation.diary10.util.NetworkStatus
import gun0912.tedkeyboardobserver.BaseKeyboardObserver
import gun0912.tedkeyboardobserver.TedKeyboardObserver
import io.reactivex.Observable
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.util.*
import kotlin.collections.HashMap

class EditAccountActivity : AppCompatActivity(), View.OnClickListener {
    private lateinit var binding: ActivityEditAccountBinding
    private val strInstance: () -> FirebaseStorage = { FirebaseStorage.getInstance() }
    private val dbInstance: () -> FirebaseDatabase = { FirebaseDatabase.getInstance() }
    private val authInstance: () -> FirebaseAuth = { FirebaseAuth.getInstance() }
    private val uid: () -> String = { authInstance().currentUser!!.uid }
    private val netStat: () -> Int = { NetworkStatus.getConnectivityStatus(this) }
    private var uriOptional: Optional<Uri> = Optional.empty()
    private lateinit var imageUrl: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_edit_account)
        FontUtil.setGlobalFont(binding.root)

        val accountImageUrl = intent!!.getStringExtra(URL_KEY)
        val name = intent!!.getStringExtra(NAME_KEY)
        val comment = intent!!.getStringExtra(COMMENT_KEY)

        Glide.with(baseContext)
                .load(accountImageUrl)
                .apply(RequestOptions().circleCrop())
                .into(binding.editAccountActivityProfileImageView)
        binding.editAccountActivityNameEditText.setText(name)
        binding.editAccountActivityCommentEditText.setText(comment)

        val clickableViewList = listOf(binding.editAccountActivityProfileImageView, binding.editAccountActivityCloseButton, binding.editAccountActivitySaveButton)
        Observable.fromIterable(clickableViewList).subscribe { it.setOnClickListener(this) }.dispose()
        TedKeyboardObserver(this).listen(object : BaseKeyboardObserver.OnKeyboardListener{
            override fun onKeyboardChange(isShow: Boolean) {
                when{
                    !isShow && binding.editAccountActivityNameEditText.isFocused -> binding.editAccountActivityNameEditText.clearFocus()
                    !isShow && binding.editAccountActivityCommentEditText.isFocused -> binding.editAccountActivityCommentEditText.clearFocus()
                }
            }
        })
    }

    override fun onClick(v: View) {
        when(v.id){
            R.id.editAccountActivity_closeButton -> finish()
            R.id.editAccountActivity_saveButton -> {
                if(netStat() == NetworkStatus.TYPE_CONNECTED){
                    updateUser()
                    val resultIntent = Intent().apply { putExtra(NAME_KEY, binding.editAccountActivityNameEditText.text.toString())
                            putExtra(COMMENT_KEY, binding.editAccountActivityCommentEditText.text.toString()) }
                    if(uriOptional != Optional.empty<Uri>()) updateProfileImage(resultIntent)
                    setResultAndFinish(resultIntent)
                }else Toast.makeText(baseContext, getString(R.string.network_not_connected), Toast.LENGTH_SHORT).show()
            }
            R.id.editAccountActivity_profileImageView ->
                startActivityForResult(Intent(Intent.ACTION_PICK).apply{ type = MediaStore.Images.Media.CONTENT_TYPE }, PICK_FROM_ALBUM_CODE)
        }
    }

    private fun updateUser(){
        val name = binding.editAccountActivityNameEditText.text.toString()
        if(name == "") Toast.makeText(baseContext, getString(R.string.write_name), Toast.LENGTH_SHORT).show()
        else {
            fun updateUser(map: HashMap<String, Any>) = dbInstance().reference.child(getString(R.string.fdb_users)).child(uid()).updateChildren(map)
            updateUser(HashMap<String, Any>().apply { put(getString(R.string.fdb_user_name), name) })
            updateUser(HashMap<String, Any>().apply { put(getString(R.string.fdb_comment), binding.editAccountActivityCommentEditText.text.toString()) })
            val userProfileChangeRequest = UserProfileChangeRequest.Builder().setDisplayName(name).build()
            authInstance().currentUser!!.updateProfile(userProfileChangeRequest)
        }
    }

    private fun updateProfileImage(resultIntent: Intent){
        setViewWhenUploading()
        strInstance().reference.child(getString(R.string.fstr_user_images)).child(uid()).putFile(uriOptional.get())
                .addOnSuccessListener {
                    strInstance().reference.child(getString(R.string.fstr_user_images)).child(uid()).downloadUrl
                            .addOnSuccessListener {
                                imageUrl = it.toString()
                                resultIntent.putExtra(URI_KEY, uriOptional.get())

                                val map = HashMap<String, Any>().apply { put(getString(R.string.fdb_profile_image_url), imageUrl) }
                                dbInstance().reference.child(getString(R.string.fdb_users)).child(uid()).updateChildren(map)
                                        .addOnSuccessListener {
                                            setResult(Activity.RESULT_OK, resultIntent)
                                            finish()
                                        }
                            }}
    }

    private fun setViewWhenUploading() {
        val viewList = listOf(binding.editAccountActivityProgressLayout, binding.editAccountActivityConstraintLayout, binding.editAccountActivityCloseButton, binding.editAccountActivitySaveButton,
                binding.editAccountActivityProfileImageView, binding.editAccountActivityCommentEditText, binding.editAccountActivityNameEditText)
        GlobalScope.launch {
            sequence{ yieldAll(viewList) }
                    .forEach { when(it.id){
                        R.id.editAccountActivity_progressLayout -> it.visibility = View.VISIBLE
                        else -> it.isEnabled = false
                    } }
        }
    }

    private fun setResultAndFinish(resultIntent: Intent){
        binding.editAccountActivityProgressBar.visibility = View.INVISIBLE
        resultIntent.putExtra(URI_KEY, "")
        setResult(Activity.RESULT_OK, resultIntent)
        finish()

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if(requestCode == PICK_FROM_ALBUM_CODE && resultCode == Activity.RESULT_OK){
            uriOptional = Optional.of(data!!.data!!)
            Glide.with(baseContext)
                    .load(data.data!!)
                    .apply(RequestOptions().circleCrop())
                    .into(binding.editAccountActivityProfileImageView)
        }
    }
}
